tests/behaviour/typeql/steps.rs

```diff
@@ -20,72 +20,221 @@
-        assert!(res.is_ok());
+    async fn typeql_define(context: &mut Context, step: &Step) -> TypeDBResult {
+        let parsed = parse_query(step.docstring().unwrap())?;
+        context.transaction().query().define(&parsed.to_string()).await
     }
 
     #[step(expr = "typeql define; throws exception")]
     async fn typeql_define_throws(context: &mut Context, step: &Step) {
-        let parsed = parse_query(step.docstring().unwrap());
-        if parsed.is_ok() {
-            let res = context.transaction().query().define(&parsed.unwrap().to_string()).await;
-            assert!(res.is_err());
-        }
+        assert_err!(typeql_define(context, step).await);
     }
 
     #[step(expr = "typeql define; throws exception containing {string}")]
     async fn typeql_define_throws_exception(context: &mut Context, step: &Step, exception: String) {
-        let result = async { parse_query(step.docstring().unwrap()).map_err(|error| error.to_string()) }
-            .and_then(|parsed| async move {
-                context.transaction().query().define(&parsed.to_string()).await.map_err(|error| error.to_string())
-            })
-            .await;
-        assert!(result.is_err());
-        assert!(result.unwrap_err().contains(&exception));
+        assert!(typeql_define(context, step).await.unwrap_err().to_string().contains(&exception));
+    }
+
+    #[step(expr = "typeql undefine")]
+    async fn typeql_undefine(context: &mut Context, step: &Step) -> TypeDBResult {
+        let parsed = parse_query(step.docstring().unwrap())?;
+        context.transaction().query().undefine(&parsed.to_string()).await
+    }
+
+    #[step(expr = "typeql undefine; throws exception")]
+    async fn typeql_undefine_throws(context: &mut Context, step: &Step) {
+        assert_err!(typeql_undefine(context, step).await);
     }
 
     #[step(expr = "typeql insert")]
-    async fn typeql_insert(context: &mut Context, step: &Step) {
-        let parsed = parse_query(step.docstring().unwrap());
-        assert!(parsed.is_ok());
-        let inserted = context.transaction().query().insert(&parsed.unwrap().to_string());
-        assert!(inserted.is_ok());
-        let res = inserted.unwrap().try_collect::<Vec<_>>().await;
-        assert!(res.is_ok());
+    async fn typeql_insert(context: &mut Context, step: &Step) -> TypeDBResult {
+        let parsed = parse_query(step.docstring().unwrap())?;
+        context.transaction().query().insert(&parsed.to_string())?.try_collect::<Vec<_>>().await?;
+        Ok(())
     }
 
     #[step(expr = "typeql insert; throws exception")]
     async fn typeql_insert_throws(context: &mut Context, step: &Step) {
+        assert_err!(typeql_insert(context, step).await);
+    }
+
+    #[step(expr = "typeql insert; throws exception containing {string}")]
+    async fn typeql_insert_throws_exception(context: &mut Context, step: &Step, exception: String) {
```

Comment:
if we look carefully, the interesting distinction between `xxx_throws` and `xxx_throws_exception` isn't actually the addition of "exception", its the "containing" that is the main difference! Both check for throws -- i'd call this `typeql_insert_throws_containing`, do you agree?